go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,482 @@
+# Go SDK Workflow Interceptor Updates and OpenTelemetry Tracing Support
+
+This proposal is for updates to Go workflow interceptors with the following features:
+
+* Support for and implementation of OpenTelemetry tracing
+* Support idiomatic Go interceptor design
+* Provide feature parity with Java SDK and other missing features (see https://github.com/temporalio/sdk-go/issues/529)
+* Add activity interceptors
+* Add client interceptors
+* Make inbound interceptors future-proof for argument addition
+* Support for using workflow/activity proto headers from interceptors
+  * One notable use case for this is to support validation (e.g. authz) and erroring out based on workflow/activity
+    header contents
+
+**These updates are known to NOT be backwards-compatible**
+
+## Interceptor Proposed Updates
+
+### Interceptor
+
+Currently does not exist.
+
+Proposal:
+
+```go
+type Interceptor interface {
+  ClientInterceptor
+  WorkerInterceptor
+}
+```
+
+There will be an `InterceptorBase` but embedding will not be required (unlike all other interfaces).
+
+Why: This allows one to have a single interceptor for all. See later for explanation of worker and client interceptors.
+
+### WorkerInterceptor
+
+Current:
+
+```go
+type WorkflowInterceptor interface {
+  InterceptWorkflow(/*...*/)
+}
+```
+
+Proposal:
+
+```go
+type WorkerInterceptor interface {
+  InterceptActivity(/*...*/)
+  InterceptWorkflow(/*...*/)
+
+  mustEmbedWorkerInterceptorBase()
+}
+```
+
+There will also be a `WorkerInterceptorBase` which must be embedded. See later for explanation of forced embeds and
+activity interceptors.
+
+Why: Maps to what Java does.
+
+### WorkerInterceptor.InterceptWorkflow
+
+Current:
+
+```go
+InterceptWorkflow(info *WorkflowInfo, next WorkflowInboundInterceptor) WorkflowInboundInterceptor
+```
+
+Proposal:
+
+```go
+InterceptWorkflow(ctx Context, next WorkflowInboundInterceptor) WorkflowInboundInterceptor
+```
+
+Why: Callers can use the context to get workflow info in addition to other things (e.g. loggers, metricsm, etc). Also
+using the context as the first parameter is good Go practice and allows for more contextual information to be provided
+in the future.
+
+### WorkflowXInterceptor
+
+Proposal: Change the name from `WorkflowXCallsInterceptor` to `WorkflowXInterceptor`. Why: `Calls` just adds noise (more
+accepted in Java, not Go).
+
+### WorkflowInboundInterceptor.ExecuteWorkflow
+
+Current:
+
+```go
+ExecuteWorkflow(ctx Context, workflowType string, args ...interface{}) []interface{}
+```
+
+Proposal:
+
+```go
+type WorkflowInput struct {
+  Args []interface{}
```

Comment:
Today, the arguments for this execution, due to how our converters work, has a pointer to every non-context-non-byte-slice argument instead of the actual values. I am looking into changing this.

Reply:
After much effort and consternation, I have worked through an implementation that uses real values and not pointers to real values. This will be another breaking change from how interceptors work today. Today, if you intercept `ExecuteWorkflow` with a `int` param, the `args` actually has it as `*int` which is nonintuitive and causes confusion when replicating this behavior for the `ExecuteActivity` use cases for local/testsuite activities that use similar code but do not have an argument unmarshaling step. With this implementation, arguments will be represented by their actual types assuming I can solve all test cases.